WireBackup/Sources/WireBackup/UseCases/CreateBackupUseCase.swift

```diff
@@ -0,0 +1,173 @@
+        userStore: UserStore,
+        conversationStore: ConversationStore,
+        messageStore: MessageStore,
+        fileArchiver: FileArchiver,
+        currentDateProvider: any CurrentDateProviding,
+        logger: @escaping @autoclosure @Sendable () -> any LoggerProtocol
+    ) {
+        self.userStore = userStore
+        self.conversationStore = conversationStore
+        self.messageStore = messageStore
+        self.fileArchiver = fileArchiver
+        self.currentDateProvider = currentDateProvider
+        self.selfUserID = selfUserID
+        self.selfUserHandle = selfUserHandle
+        self.logger = logger
+    }
+
+    public func invoke(password: String) -> AsyncThrowingStream<CreateBackupProgress, any Error> {
+        AsyncThrowingStream { continuation in
+            let task = Task<Void, Never> { [
+                // swiftlint:disable closure_parameter_position
+                userStore,
+                conversationStore,
+                messageStore,
+                currentDateProvider,
+                fileArchiver,
+                logger,
+                selfUserID,
+                selfUserHandle
+                // swiftlint:enable closure_parameter_position
+            ] in
+
+                let workDirectoryURL = URL(fileURLWithPath: NSTemporaryDirectory())
+                    .appendingPathComponent(UUID().uuidString)
+                let outputDirectoryURL = URL(fileURLWithPath: NSTemporaryDirectory())
+                    .appendingPathComponent(UUID().uuidString)
+
+                let fileManager = FileManager.default
+                defer { try? fileManager.removeItem(at: workDirectoryURL) }
+
+                do {
+                    let logger = logger()
+                    let reportProgress: (Int, Int) -> Void = { current, total in
+                        guard current % 50 == 0 || current == total else { return }
+                        logger.debug("reporting overall process: \(current)/\(total)")
+                        continuation.yield(.progress(current, total))
+                    }
+
+                    reportProgress(0, 0)
+
+                    logger.debug("initializing backup creator")
+                    let backupCreator = BackupCreator(
+                        selfUserID: selfUserID,
+                        workDirectoryURL: workDirectoryURL,
+                        outputDirectoryURL: outputDirectoryURL,
+                        fileArchiver: fileArchiver
+                    )
+
+                    try Task.checkCancellation()
+
+                    // get the counts of users, messages and conversations in order to report progress accurately
+                    logger.debug("calculating entity counts")
+                    let userCount = try await userStore.totalUserCount()
+                    let conversationCount = try await conversationStore.totalConversationCount()
+                    let messageCount = try await messageStore.totalMessageCount()
```

Comment:
suggestion: I think having a single specific store that exposes the api necessary for the current task at hand may give the benefit of fewer context switches. For instance:

```swift
protocol BackupLocalStore {

    func countBackupModels() async throws -> (users: Int, conversations: Int, messages: Int)
    func fetchAllUsers() async throws -> Set<UserBackupModel>
    func fetchAllConversations() async throws -> Set<ConversationBackupModel>
    func fetchAllMessages() async throws -> Set<MessageBackupModel>
    ...

}
```

Reply:
Done, thanks.